plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php

```diff
@@ -0,0 +1,252 @@
+		}
+	}
+
+	/**
+	 * Validates and saves additional address fields to the customer object on the My Account page.
+	 *
+	 * @param integer $user_id User ID.
+	 */
+	public function save_account_form_fields( $user_id ) {
+		$customer = new WC_Customer( $user_id );
+		$fields   = $this->checkout_fields_controller->get_fields_for_location( 'contact' );
+
+		foreach ( $fields as $field_key => $field ) {
+			if ( ! isset( $_POST[ $field_key ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
+				continue;
+			}
+			$this->checkout_fields_controller->persist_field_for_customer( $field_key, wc_clean( wp_unslash( $_POST[ $field_key ] ) ), $customer );
+		}
+	}
+
+	/**
+	 * Adds additional address fields to the My Account edit address form.
+	 *
+	 * @param array  $address Address fields.
+	 * @param string $address_type Type of address (billing or shipping).
+	 * @return array Updated address fields.
+	 */
+	public function edit_address_fields( $address, $address_type ) {
+		$customer = new WC_Customer( get_current_user_id() );
+		$fields   = $this->checkout_fields_controller->get_fields_for_location( 'address' );
+
+		foreach ( $fields as $key => $field ) {
+			$field_key                      = 'billing' === $address_type ? '/billing/' . $key : '/shipping/' . $key;
+			$address[ $field_key ]          = $field;
+			$address[ $field_key ]['value'] = $this->checkout_fields_controller->get_field_from_customer( $key, $customer, $address_type );
+
+			if ( 'select' === $field['type'] ) {
+				$address[ $field_key ]['options'] = array_column( $field['options'], 'label', 'value' );
+			}
+
+			if ( 'checkbox' === $field['type'] ) {
+				$address[ $field_key ]['checked_value']   = '1';
+				$address[ $field_key ]['unchecked_value'] = '0';
+			}
+		}
+
+		return $address;
+	}
+
+	/**
+	 * Validates and saves additional address fields to the customer object on the My Account page.
+	 *
+	 * @param integer $user_id User ID.
+	 * @param string  $address_type Type of address (billing or shipping).
+	 */
+	public function save_address_fields( $user_id, $address_type ) {
+		$customer = new WC_Customer( $user_id );
+		$fields   = $this->checkout_fields_controller->get_fields_for_location( 'address' );
+
+		foreach ( $fields as $key => $field ) {
+			$field_key = 'billing' === $address_type ? '/billing/' . $key : '/shipping/' . $key;
+
+			if ( ! isset( $_POST[ $field_key ] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.Missing
+				continue;
+			}
+
+			// phpcs:ignore WordPress.Security.NonceVerification.Missing
+			$value = wc_clean( wp_unslash( $_POST[ $field_key ] ) );
+
+			if ( ! empty( $field['required'] ) && empty( $value ) ) {
+				// translators: %s field label.
+				wc_add_notice( sprintf( __( '%s is a required field.', 'woocommerce' ), $field['label'] ), 'error' );
+				continue;
+			}
+
+			$this->checkout_fields_controller->persist_field_for_customer( $field_key, $value, $customer );
```

Comment:
Should this also be validating it against any validation hooks in place? e.g. add this code:

```php

woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/gov-id',
				'label'    => 'Government ID',
				'location' => 'address',
				'type'     => 'text',
				'required' => true,
			),
		);

add_filter('woocommerce_blocks_validate_additional_field_plugin-namespace/gov-id', function ( \WP_Error $error, $value, $schema, $request ) {
			$match = preg_match( '/[A-Z0-9]{5}/', $value );
			if ( 0 === $match || false === $match ) {
				$error->add( 'invalid_gov_id', 'Please ensure your government ID matches the correct format.' );
			}
			return $error;
		}, 10, 4);
```

then save your address with an invalid format (e.g. `abcde`) I would expect this not to save. Same for contact

Reply:
Could we tackle this in a follow up? The current validation method requires schema so I'm not certain how to consume that from these classes. If we have a method we can call, we can add a validation_callback when registering fields and add support in woo core to check it.

Reply:
Yeah, makes sense to do this as a follow up, any further orders with this info will be re-validated anyway

Reply:
Added an issue: https://github.com/woocommerce/woocommerce/issues/44019